lib/src/query/filter.dart

```
@@ -1,35 +1,87 @@
-    ..removeWhere((k, v) => k.isEmpty));
+  static Filter fromUri(Uri uri) {
+    Map<String, Object> filters = {};
+    uri.queryParametersAll.forEach((key, value) {
+      if (_validationRegex.hasMatch(key)) {
+        final matches = _extractionRegex.allMatches(key).toList();
+        _convertToMapAndMerge(matches, filters, value.last);
+      }
+    });
+    return Filter(filters);
+  }
+
+  static void _convertToMapAndMerge(
+    List<RegExpMatch> matches,
+    Map<String, Object> destination,
+    String value,
+  ) {
+    final key = matches[0].group(1) ?? '';
+    if (key.isNotEmpty) {
+      if (matches.length == 1) {
+        destination[key] = value;
+        return;
+      }
+      if (!destination.containsKey(key) ||
+          destination[key] is! Map<String, Object>) {
+        destination[key] = <String, Object>{};
+      }
+      _convertToMapAndMerge(
+        matches.sublist(1),
+        destination[key] as Map<String, Object>,
+        value,
+      );
+    }
+  }
 
-  static final _regex = RegExp(r'^filter\[(.+)\]$');
+  static final _validationRegex = RegExp(r'^filter(?:\[[^\[\]]+\])+$');
+  static final _extractionRegex = RegExp(r'\[([^\[\]]+)\]');
 
-  final _ = <String, String>{};
+  final _ = <String, Object>{};
 
   /// Converts to a map of query parameters
-  Map<String, String> get asQueryParameters =>
-      _.map((k, v) => MapEntry('filter[$k]', v));
+  Map<String, String> get asQueryParameters => _flattenFiltersMap(_);
+
+  Map<String, String> _flattenFiltersMap(Map<String, Object> filters,
+      [String keyPrefix = 'filter']) {
+    final queryParameters = <String, String>{};
+    filters.forEach((key, value) {
+      final keyWithPrefix = '$keyPrefix[$key]';
+      if (value is String) {
+        queryParameters[keyWithPrefix] = value;
+      } else if (value is Map<String, Object>) {
+        queryParameters.addAll(_flattenFiltersMap(value, keyWithPrefix));
+      }
```

Would an else block throwing an argument error be useful?

---

Nothing bad will happen if the value is not one of the types above, it will just be ignored. With that said the error might be a good idea as it could be helpful to the developer, so that the filters with wrong type won't be just silently ignored.

---

Yeah that's what I figured too.